### PR TITLE
Use Caching in GitHub Actions

### PR DESCRIPTION
**Pull request by MH321Productions**
# Readiness checklist

-   [ ] I added/updated unit tests.
-   [ ] I added/updated integration tests.
-   [x] I ensured that the PR title is good enough for the changelog.
-   [x] I labeled the PR.

The caching cuts the runtime of the workflow by up to 4 minutes (~11 minutes to ~7 minutes)

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -53,28 +73,168 @@ jobs:
...
+          - dockerfile: ConsumerApi/Dockerfile
+            tag: consumer-api
+          - dockerfile: AdminApi/src/AdminApi/Dockerfile
+            tag: admin-ui
+          - dockerfile: Modules/Devices/src/Devices.AdminCli/Dockerfile
+            tag: seed-client
+          - dockerfile: DatabaseMigrator/Dockerfile
+            tag: database-migrator
+          - dockerfile: EventHandlerService/src/EventHandlerService/Dockerfile
+            tag: event-handler-service
```
Comment (tnotheis) [inline .github/workflows/test.yml:122]: Why do we have tags that are named just like the image? If the tag is mandatory, we can just hardcode it to something like "test"

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -53,28 +73,168 @@ jobs:
...
+        run: docker compose -f ./.ci/docker-compose.test.yml -f ./.ci/docker-compose.test.postgres.yml logs > docker-log.txt
+
+      - name: Archive logs
+        if: failure()
+        uses: actions/upload-artifact@v4
+        with:
+          name: integration-test-postgres-docker-logs
+          path: docker-log.txt
+        env:
+          DOTNET_CONSOLE_ANSI_COLOR: true
```
Comment (tnotheis) [inline .github/workflows/test.yml:183]: Why are you saving docker compose logs when you didn't even do anything with compose in this job?

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -53,28 +73,168 @@ jobs:
...
+          path: ~/.nuget/packages
+          key: ${{ runner.os }}-nuget-${{ hashFiles('**/*.csproj') }}
+          restore-keys: ${{ runner.os }}-nuget-
       - name: Run tests
         run: ./.ci/test.sh
         env:
           DOTNET_CONSOLE_ANSI_COLOR: true
+          
+  container-test-builds:
+    name: Build ${{matrix.tag}} Container
```
Comment (tnotheis) [inline .github/workflows/test.yml:108]: You only build the image. Not the container.

File: .github/workflows/test.yml (hunk 1 above)
Comment (tnotheis) [inline .github/workflows/test.yml:122]: Is there a specific reason why you only build these five images instead of all?

Also, you should be able to remove the jobs that are specifically for building the images. That's just duplication, I think.

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -87,16 +247,48 @@ jobs:
...
+          pattern: docker-*
+          merge-multiple: true
+
+      - name: Load Docker images
+        run: |
+          docker load --input /tmp/consumer-api.tar
+          docker load --input /tmp/admin-ui.tar
+          docker load --input /tmp/seed-client.tar
+          docker load --input /tmp/database-migrator.tar
+          docker load --input /tmp/event-handler-service.tar
```
Comment (tnotheis) [inline .github/workflows/test.yml:283]: This seems to be very error prone. You can easily forget to add a new image here. And then you're running without cache

Isn't it possible to just load all tars?

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -87,16 +247,48 @@ jobs:
...
+
+      - name: Setup NuGet
+        uses: actions/setup-dotnet@v4
+        with:
+          dotnet-version: 8.x
+
       - name: Install script dependencies
         run: npm install --prefix ./.ci
+
+      - name: Download Docker images
```
Comment (tnotheis) [inline .github/workflows/test.yml:270]:
```suggestion
      - name: Download cached Docker images
```

Same for the other occurrences

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: .ci/integrationTest.postgres.sh [hunk 6]
```diff
@@ -5,11 +5,17 @@ dockerCompose() {
...
+  dotnet restore "Backbone.sln";
+  dotnet build --no-restore "Backbone.sln"
+}
+wait
+
 export CONSUMER_API_BASE_ADDRESS="http://localhost:5000"
 export ADMIN_API_BASE_ADDRESS="http://localhost:5173"
-dotnet test --no-restore --no-build --filter "Category=Integration&TestCategory!~ignore" --logger "GitHubActions;summary.includePassedTests=true" "Backbone.sln"
+
+dotnet test --no-restore --no-build --filter "Category=Integration&TestCategory!~ignore" --logger "GitHubActions;summary.includePassedTests=true;summary.includeSkippedTests=true" "Backbone.sln"
```
[the remark this answers is not in the document]
Comment (MH321Productions) [reply .ci/integrationTest.postgres.sh:21]: I just copied the arguments from another place as part of the usage of the GitHub Actions logger

File: .github/workflows/test.yml (hunk 1 above)
Comment (MH321Productions) [reply .github/workflows/test.yml:122]: These are the only images containing a build instruction in the compose files (as this was my focus)

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -38,12 +38,32 @@ jobs:
     steps:
       - name: Checkout
         uses: actions/checkout@v4
+      - uses: actions/setup-node@v3
+        with:
+          node-version: 20
+          check-latest: true
```
[the remark this answers is not in the document]
Comment (tnotheis) [reply .github/workflows/test.yml:44]: Yes.

File: .ci/integrationTest.postgres.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (tnotheis) [reply .ci/integrationTest.postgres.sh:21]: But not from our code, right? Because we don't use `summary.includeSkippedTests=true` anywhere, as far as I can see.

File: .github/workflows/test.yml (hunk 1 above)
Comment (tnotheis) [reply .github/workflows/test.yml:122]: Yeah, I was expecting that. But I like your solution for two reasons:
1. It leads to less code
2. It then also benefits from the caching of the used base images.

File: .ci/integrationTest.postgres.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (MH321Productions) [reply .ci/integrationTest.postgres.sh:21]: I don't know anymore. I just removed it

File: .github/workflows/test.yml (hunk 1 above)
Comment (tnotheis) [reply .github/workflows/test.yml]: Why did you ignore this comment?

Comment (MH321Productions) [reply .github/workflows/test.yml]: I didn't ignore it, I had to work on the release v5 PR, but thought about a solution

File: .ci/integrationTest.postgres.sh [hunk 8]
```diff
@@ -5,11 +5,17 @@ dockerCompose() {
...
+  dotnet restore "Backbone.sln";
+  dotnet build --no-restore "Backbone.sln"
+}
+wait
+
 export CONSUMER_API_BASE_ADDRESS="http://localhost:5000"
 export ADMIN_API_BASE_ADDRESS="http://localhost:5173"
-dotnet test --no-restore --no-build --filter "Category=Integration&TestCategory!~ignore" --logger "GitHubActions;summary.includeNotFoundTests=false" "Backbone.sln"
+
+dotnet test --no-restore --no-build --filter "Category=Integration&TestCategory!~ignore" --logger "GitHubActions;summary.includePassedTests=true;summary.includeNotFoundTests=false" "Backbone.sln"
```
Comment (tnotheis) [inline .ci/integrationTest.postgres.sh:21]: Why did you add `summary.includePassedTests=true`?

File: .ci/integrationTest.sqlserver.sh [hunk 9]
```diff
@@ -5,11 +5,17 @@ dockerCompose() {
...
+  dotnet restore "Backbone.sln";
+  dotnet build --no-restore "Backbone.sln"
+}
+wait
+
 export CONSUMER_API_BASE_ADDRESS="http://localhost:5000"
 export ADMIN_API_BASE_ADDRESS="http://localhost:5173"
-dotnet test --no-restore --no-build --filter "Category=Integration&TestCategory!~ignore" --logger "GitHubActions;summary.includeNotFoundTests=false" "Backbone.sln"
+
+dotnet test --no-restore --no-build --filter "Category=Integration&TestCategory!~ignore" --logger "GitHubActions;summary.includePassedTests=true;summary.includeNotFoundTests=false" "Backbone.sln"
```
Comment (tnotheis) [inline .ci/integrationTest.sqlserver.sh:21]: See above

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: .ci/integrationTest.postgres.sh (hunk 8 above)
Comment (MH321Productions) [reply .ci/integrationTest.postgres.sh:21]: I'm not sure what was copied in and what was merged in

**Review by tnotheis — APPROVED** [review]
